test/unit/endpoint-client.test.ts

```diff
@@ -1,307 +1,501 @@
+	beforeEach(() => {
+		client = buildClient()
 	})
 
 	afterEach(() => {
 		jest.clearAllMocks()
 	})
 
-	test('request', async () => {
-		const response = await client.request('GET', 'mypath')
-		expect(axios.request).toHaveBeenCalledTimes(1)
-		expect(axios.request).toHaveBeenCalledWith({
-			'url': 'https://api.smartthings.com/basepath/mypath',
-			'method': 'GET',
-			'headers': {
-				'Content-Type': 'application/json;charset=utf-8',
-				'Accept': 'application/json',
-				'Authorization': `Bearer ${token}`,
-				'X-ST-CORRELATION': 'AAABBBCCC',
-			},
-			'data': undefined,
-			'params': undefined,
-			'paramsSerializer': expect.anything(),
+	describe('setHeader', () => {
+		it('adds header to config', () => {
+			client.setHeader('NewHeader', 'header value')
+
+			expect(client.config.headers?.NewHeader).toBe('header value')
+		})
+
+		it('works when no previous headers set', () => {
+			const client = new EndpointClient('base/path', { ...configWithoutHeaders })
+			client.setHeader('NewHeader', 'header value')
+
+			expect(client.config.headers?.NewHeader).toBe('header value')
 		})
-		expect(response.status).toBe('ok')
 	})
 
-	test('request with header overrides', async () => {
-		const headerOverrides = {
-			'Content-Type': 'overridden content type',
-			'X-ST-Organization': '00000000-0000-0000-0000-000000000008',
-		}
-		const response = await client.request('POST', 'mypath', { name: 'Bob' }, undefined, { headerOverrides })
-		expect(axios.request).toHaveBeenCalledTimes(1)
-		expect(axios.request).toHaveBeenCalledWith({
-			'url': 'https://api.smartthings.com/basepath/mypath',
-			'method': 'POST',
-			'headers': {
+	describe('removeHeader', () => {
+		it('removes header from config', () => {
+			client.setHeader('NewHeader', 'header value')
+
+			expect(client.config.headers?.NewHeader).toBe('header value')
+
+			client.removeHeader('NewHeader')
+
+			expect(client.config.headers?.NewHeader).toBeUndefined()
+		})
+
+		it('ignores undefined headers', () => {
+			client.removeHeader('NewHeader')
+
+			expect(client.config.headers?.NewHeader).toBeUndefined()
+		})
+	})
+
+	describe('request', () => {
+		it('submits basic request', async () => {
+			const response = await client.request('GET', 'my/path')
+
+			expect(mockRequest).toHaveBeenCalledTimes(1)
+			expect(mockRequest).toHaveBeenCalledWith({
+				url: 'https://api.smartthings.com/base/path/my/path',
+				method: 'GET',
+				headers: {
+					...headers,
+					Authorization: `Bearer ${token}`,
+				},
+				data: undefined,
+				params: undefined,
+				paramsSerializer: expect.any(Function),
+			})
+			expect(response.status).toBe('ok')
+
+			const stringifyMock = (qs.stringify as jest.Mock<string, [unknown]>)
+				.mockReturnValue('stringified parameters')
+			const paramsSerializer = mockRequest.mock.calls[0][0].paramsSerializer
+
+			if (!paramsSerializer) {
+				// using expect doesn't satisfy TypeScript compiler
```

Comment:
Does asserting the type on `paramsSerializer` and then using [.toBeDefined()](https://jestjs.io/docs/expect#tobedefined) work?

Reply:
That does. I had tried the `toBeDefined` but without the type assertion before since VS Code tells me the type correctly on hover (which usually means TypeScript can tell it).